File: rust/2015/day1.rs
use std::fs::read_to_string;

fn main() {
    let input = read_to_string("inputs/2015/day1.data")
                .expect("unable to read file");
    let iter = input.trim().split("");

    let mut floor = 0;
    let mut basement_instruction_number: Option<i32> = None;

    for (instruction_number, instruction) in iter.enumerate() {
        floor += match instruction {
            "(" => 1,
            ")" => -1,
            _ => 0,
        };
        if floor == -1 && basement_instruction_number == None {
            basement_instruction_number = Some(instruction_number as i32);
        };
    }

    let convert_to_string = |input: Option<i32>| {
        match input {
            Some(x) => x.to_string(),
            None => "Santa never goes to the basement".to_string(),
        }
    };

    println!("part 1 answer: {}", floor);
    println!("part 2 answer: {}", convert_to_string(basement_instruction_number))
}

File: python/2015/day1.py

def count_floors(input: str):
    floor = 0 
    for character in input:
        if character == '(':
            floor += 1
        elif character == ')':
            floor -=1
    return floor

def find_basement(input: str):
    floor = 0
    instruction = 0
    for character in input:
        instruction +=1
        if character == '(':
            floor += 1
        elif character == ')':
            floor -=1
        if floor == -1:
            break
    return instruction


def part1():
    print("\n~~~ Part 1 ~~~\n")
    with open("input.data", "r") as input_file:
        print(count_floors(input_file.read()))


def part2():
    print("\n~~~ Part 2 ~~~\n")
    with open("input.data", "r") as input_file:
        print(find_basement(input_file.read()))


if __name__ == "__main__":
    part1()
    part2()